src/main/scala/beam/router/skim/ODSkimmer.scala

```diff
@@ -0,0 +1,673 @@
+  private val observedSkimsFileBaseName = "odskims"
+  private val fullSkimsFileBaseName = "odskimsFull"
+  private val excerptSkimsFileBaseName = "odskimsExcerpt"
+
+  override val aggregatedSkimsFilePath: String = beamConfig.beam.warmStart.skimsFilePath
+
+  override def handleEvent(event: Event): Unit = {
+    event match {
+      case e: ODSkimmerEvent =>
+        observeTrip(e.trip, e.generalizedTimeInHours, e.generalizedCost, e.energyConsumption)
+      case _ =>
+    }
+  }
+
+  override def writeToDisk(event: IterationEndsEvent): Unit = {
+    if (beamConfig.beam.skimmanager.odskimmer.writeObservedSkimsInterval > 0 && event.getIteration % beamConfig.beam.skimmanager.odskimmer.writeObservedSkimsInterval == 0) {
+      ProfilingUtils.timed(s"writeObservedSkims on iteration ${event.getIteration}", x => logger.info(x)) {
+        writeObservedSkims(event)
+      }
+    }
+    if (beamConfig.beam.skimmanager.odskimmer.writeAllModeSkimsForPeakNonPeakPeriodsInterval > 0 && event.getIteration % beamConfig.beam.skimmanager.odskimmer.writeAllModeSkimsForPeakNonPeakPeriodsInterval == 0) {
+      ProfilingUtils.timed(
+        s"writeAllModeSkimsForPeakNonPeakPeriods on iteration ${event.getIteration}",
+        x => logger.info(x)
+      ) {
+        writeAllModeSkimsForPeakNonPeakPeriods(event)
+      }
+    }
+    if (beamConfig.beam.skimmanager.odskimmer.writeFullSkimsInterval > 0 && event.getIteration % beamConfig.beam.skimmanager.odskimmer.writeFullSkimsInterval == 0) {
+      ProfilingUtils.timed(s"writeFullSkims on iteration ${event.getIteration}", x => logger.info(x)) {
+        writeFullSkims(event)
+      }
+    }
+    if (beamConfig.beam.skimmanager.odskimmer.writeAggregatedSkimsInterval > 0 && event.getIteration % beamConfig.beam.skimmanager.odskimmer.writeAggregatedSkimsInterval == 0) {
+      ProfilingUtils.timed(s"writeFullSkims on iteration ${event.getIteration}", x => logger.info(x)) {
+        writeAggregatedSkims(event)
+      }
+    }
+  }
+
+  override def fromCsv(
+    line: immutable.Map[String, String]
+  ): immutable.Map[AbstractSkimmerKey, AbstractSkimmerInternal] = {
+    immutable.Map(
+      ODSkimmerKey(
+        line("timeBin").toInt,
+        BeamMode.fromString("mode").get,
+        Id.create(line("idTaz"), classOf[TAZ]),
+        Id.create(line("idTaz"), classOf[TAZ])
+      )
+      -> ODSkimmerInternal(
+        line("travelTimeInS").toDouble,
+        line("generalizedTimeInS").toDouble,
+        line("generalizedCost").toDouble,
+        line("distanceInM").toDouble,
+        line("cost").toDouble,
+        line("numObservations").toInt,
+        Option(line("energy")).map(_.toDouble).getOrElse(0.0)
+      )
+    )
+  }
+
+  override protected def getPastSkims: List[Map[AbstractSkimmerKey, AbstractSkimmerInternal]] = {
+    ODSkimmer.pastSkims.map(
+      list => list.map(kv => kv._1.asInstanceOf[AbstractSkimmerKey] -> kv._2.asInstanceOf[AbstractSkimmerInternal])
```

Comment:
Same as elsewhere - and for most of the below:

```suggestion
      list => list.map{case (key, value) => key.asInstanceOf[AbstractSkimmerKey] -> value.asInstanceOf[AbstractSkimmerInternal])
```